tutorials/pubsub-spring-dedup-messages/index.md

```diff
@@ -0,0 +1,244 @@
+
+[embedmd]:# (pubsub-spring/src/main/java/com/google/example/App.java java /  \/\/ The Consumer Bean/ /}/)
+```java
+  // The Consumer Bean makes the function a valid Spring Cloud Stream sink. It
+  // receives messages from a Cloud Pub/Sub subscription configured with the binding
+  // name `receiveDedupedMessagesFromDataflow-in-0` in application.properties.
+  @Bean
+  Consumer<Message<String>> receiveDedupedMessagesFromDataflow() {
+    return msg -> {
+      System.out.println("\tDE-DUPED message: \"" + msg.getPayload() + "\".");
+    }
+```
+
+### Specify a Cloud Pub/Sub topic and subscription for the sink
+Spring can find the Cloud Pub/Sub subscription that your code should receive data from when you provide a subscription name in `application.properties` and assign it to a consumer group of the sink using an input binder. Here, the input binding name is `receiveDedupedMessagesFromDataflow-in-0`. Because a subscription cannot exist without a topic, a topic must also be specified for the sink. Note that only input bindings have consumer groups. For more information, see [Common Binding Properties].
+
+In [application.properties](pubsub-spring/src/main/resources/application.properties):
+
+[embedmd]:# (pubsub-spring/src/main/resources/application.properties /.*Data coming/ /=subscriptionFromDataflow/)
+```properties
+# Data coming from Cloud Pub/Sub to a Spring Cloud Stream sink defined in the
+# functional bean `receiveDedupedMessagesFromDataflow`. The application
+# will create the Cloud Pub/Sub topic `topicFromDataflow` and subscription
+# `topicFromDataflow.subscriptionFromDataflow` if they do not exist.
+spring.cloud.stream.bindings.receiveDedupedMessagesFromDataflow-in-0.destination=topicFromDataflow
+spring.cloud.stream.bindings.receiveDedupedMessagesFromDataflow-in-0.group=subscriptionFromDataflow
+```
+### Run the application
+To start your application, navigate to `pubsub-spring/` and run: 
+
+```shell script
+mvn spring-boot:run
+```
+
+Observe that your app has started successfully by pointing your browser to `localhost:8080`. You should be able to send messages using the form there. At this point, Spring will have automatically created the Cloud Pub/Sub topic that you have specified in `application.properties` to publish your message to. You can view Publish Message Request Count and Publish Message Operation Count in the topic details in [Cloud Console for Pub/Sub Topic] to verify that publishing to Cloud Pub/Sub is successful.  
+
+## Start a Cloud Dataflow Job to Deduplicate Pub/Sub Messages
+
+As a middle process that takes data from a Cloud Pub/Sub topic and publishes processed data to another Cloud Pub/Sub topic, Cloud Dataflow achieves exactly once stream processing by asking the input stream for an `idAttribute`. Using the user-defined key name as this unique record identifier name, Cloud Dataflow can avoid processing messages of the same key multiple times.
```

Comment:
This is a little confusing. Is `idAttribute` the unique record identifier name? If so, what is the "user-defined key name"?